### PR TITLE
Split vertical horizontal extent

### PR DESCRIPTION
**Pull request by mroethlin**
## Technical Description

This PR splits the Extents into horizontal and vertical extents.

### Dependencies

This PR depends on the split of the _Offsets_ into vertical and horizontal offsets.

### CODE REVIEW

File: dawn/src/dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.cpp [hunk 1]
```diff
@@ -29,16 +29,16 @@ namespace dawn {
...
 namespace cxxnaive {
 
-static std::string makeLoopImpl(const iir::Extent extent, const std::string& dim,
+static std::string makeLoopImpl(const std::pair<int, int>& extent, const std::string& dim,
                                 const std::string& lower, const std::string& upper,
                                 const std::string& comparison, const std::string& increment) {
-  return Twine("for(int " + dim + " = " + lower + "+" + std::to_string(extent.Minus) + "; " + dim +
-               " " + comparison + " " + upper + "+" + std::to_string(extent.Plus) + "; " +
-               increment + dim + ")")
+  return Twine("for(int " + dim + " = " + lower + "+" + std::to_string(std::get<0>(extent)) + "; " +
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.cpp]: how about
```
auto [iextent, jextent] = extent;
return ... + std::to_string(iextent) + ... + std::to_string(jextent) + ...
```
to name the extents?

File: dawn/src/dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.cpp [hunk 2]
```diff
@@ -58,8 +58,8 @@ static std::string makeKLoop(const std::string dom, bool isBackward,
   const std::string lower = makeIntervalBound(dom, interval, iir::Interval::Bound::lower);
   const std::string upper = makeIntervalBound(dom, interval, iir::Interval::Bound::upper);
 
-  return isBackward ? makeLoopImpl(iir::Extent{}, "k", upper, lower, ">=", "--")
-                    : makeLoopImpl(iir::Extent{}, "k", lower, upper, "<=", "++");
+  return isBackward ? makeLoopImpl(std::pair<int, int>(0, 0), "k", upper, lower, ">=", "--")
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.cpp]: std::make_pair(0, 0)

File: dawn/src/dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.cpp [hunk 3]
```diff
@@ -381,10 +381,17 @@ void CXXNaiveCodeGen::generateStencilClasses(
               for(const auto& stagePtr : multiStage.getChildren()) {
                 const iir::Stage& stage = *stagePtr;
 
+                auto extents = dawn::iir::extent_cast<dawn::iir::CartesianExtent const&>(
+                    stage.getExtents().horizontalExtent());
+
                 stencilRunMethod.addBlockStatement(
-                    makeIJLoop(stage.getExtents()[0], "m_dom", "i"), [&]() {
+                    makeIJLoop(std::pair<int, int>(extents.iMinus(), extents.iPlus()), "m_dom",
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/CXXNaive/CXXNaiveCodeGen.cpp]: make_pair

File: dawn/src/dawn/CodeGen/Cuda/CacheProperties.cpp [hunk 4]
```diff
@@ -27,26 +27,44 @@ makeCacheProperties(const std::unique_ptr<iir::MultiStage>& ms,
                     const std::shared_ptr<iir::StencilInstantiation>& stencilInstantiation,
                     const int maxRedundantLines) {
 
-  iir::Extents maxExtents{0, 0, 0, 0, 0, 0};
+  iir::Extents maxExtents(dawn::ast::cartesian_{}, 0, 0, 0, 0, 0, 0);
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/Cuda/CacheProperties.cpp]: here and everywhere else, you can use `dawn::ast::cartesian` instead of `dawn::ast::cartesian_{}`

File: dawn/src/dawn/CodeGen/Cuda/CacheProperties.cpp [hunk 5]
```diff
@@ -27,26 +27,44 @@ makeCacheProperties(const std::unique_ptr<iir::MultiStage>& ms,
...
-    auto extents = ms->getField(accessID).getExtentsRB();
+    auto vExtents = ms->getField(accessID).getExtentsRB().verticalExtent();
+    auto hExtents = dawn::iir::extent_cast<dawn::iir::CartesianExtent const&>(
+        ms->getField(accessID).getExtentsRB().horizontalExtent());
+
     bool exceeds = false;
-    for(int i = 0; i < 3; ++i) {
-      if(extents[i].Minus < -maxRedundantLines || extents[i].Plus > maxRedundantLines) {
+
+    const int numDimensions = 3;
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/Cuda/CacheProperties.cpp]: `constexpr`

File: dawn/src/dawn/CodeGen/Cuda/CacheProperties.cpp [hunk 6]
```diff
@@ -132,30 +150,48 @@ iir::Extents CacheProperties::getCacheExtent(int accessID) const {
...
 int CacheProperties::getStrideCommonCache(int dim, Array3ui blockSize) const {
-  return getStrideImpl(dim, blockSize, extents_);
-}
+  auto hExtents =
+      dawn::iir::extent_cast<dawn::iir::CartesianExtent const&>(extents_.horizontalExtent());
+  return getStrideImpl(dim, blockSize, hExtents.iMinus(), hExtents.iPlus());
+} // namespace cuda
 
-int CacheProperties::getStrideImpl(int dim, Array3ui blockSize, const iir::Extents& extents) const {
+int CacheProperties::getStrideImpl(int dim, Array3ui blockSize, int Minus, int Plus) const {
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/Cuda/CacheProperties.cpp]: name convention: lower case `minus`  / `plus`

File: dawn/src/dawn/CodeGen/Cuda/MSCodeGen.cpp [hunk 7]
```diff
@@ -310,7 +317,7 @@ void MSCodeGen::generateFillKCaches(MemberFunction& cudaKernel, const iir::Inter
       continue;
     }
 
-    auto intervalVertExtent = (*extents)[2];
+    auto intervalVertExtent = (*extents).verticalExtent();
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/Cuda/MSCodeGen.cpp]: `extents->verticalExtent`

File: dawn/src/dawn/CodeGen/Cuda/MSCodeGen.cpp [hunk 8]
```diff
@@ -459,11 +467,11 @@ MSCodeGen::buildKCacheProperties(const iir::Interval& interval,
 
     if(applyEpflush) {
       auto cacheName = cacheProperties_.getCacheName(accessID);
-      auto intervalVertExtent = (*extents)[2];
+      auto intervalVertExtent = (*extents).verticalExtent();
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/Cuda/MSCodeGen.cpp]: extents->verticalExtent()

File: dawn/src/dawn/CodeGen/StencilFunctionAsBCGenerator.cpp [hunk 9]
```diff
@@ -47,13 +47,15 @@ void StencilFunctionAsBCGenerator::visit(const std::shared_ptr<iir::VarAccessExp
...
-  int haloIMinus = abs(extents[0].Minus);
-  int haloIPlus = abs(extents[0].Plus);
-  int haloJMinus = abs(extents[1].Minus);
-  int haloJPlus = abs(extents[1].Plus);
-  int haloKMinus = abs(extents[2].Minus);
-  int haloKPlus = abs(extents[2].Plus);
+  const auto& hExtents = dawn::iir::extent_cast<dawn::iir::CartesianExtent const&>(
+      metadata_.getBoundaryConditionExtentsFromBCStmt(stmt).horizontalExtent());
+  const auto& vExtents = metadata_.getBoundaryConditionExtentsFromBCStmt(stmt).verticalExtent();
+  int haloIMinus = abs(hExtents.iMinus());
```
Comment (lukasm91) [inline dawn/src/dawn/CodeGen/StencilFunctionAsBCGenerator.cpp]: can you change to `std::abs`? abs is C

File: dawn/src/dawn/IIR/ASTStmt.cpp [hunk 10]
```diff
@@ -110,7 +110,7 @@ std::optional<Extents> computeMaximumExtents(Stmt& stmt, const int accessID) {
       if(extents)
         extents->merge(*childExtent);
       else
-        extents = childExtent;
+        extents.swap(childExtent);
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/ASTStmt.cpp]: why is this needed? `operator=` should work properly. In case we need `swap`, normal `swap` should do the right thing with our move semantics.

File: dawn/src/dawn/IIR/Accesses.cpp [hunk 11]
```diff
@@ -92,15 +96,15 @@ void Accesses::mergeWriteExtent(int AccessID, const Extents& extent) {
 void Accesses::addReadExtent(int AccessID, const Extents& extent) {
   auto it = readAccesses_.find(AccessID);
   if(it != readAccesses_.end())
-    it->second.add(extent);
+    it->second = Extents::add(it->second, extent);
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/Accesses.cpp]: should we implement `operator+`?

File: dawn/src/dawn/IIR/DependencyGraphAccesses.cpp [hunk 12]
```diff
@@ -545,10 +551,18 @@ void DependencyGraphAccesses::toJSON(const std::string& file, DiagnosticsEngine&
 
   auto extentsToVec = [&](const Extents& extents) {
     std::vector<int> extentsVec;
-    for(const Extent& extent : extents.getExtents()) {
-      extentsVec.push_back(extent.Minus);
-      extentsVec.push_back(extent.Plus);
-    }
+
+    auto vExtent = extents.verticalExtent();
+    auto hExtent =
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/DependencyGraphAccesses.cpp]: this one will be critical for the unstructured extents :-(

File: dawn/src/dawn/IIR/Extents.cpp [hunk 13]
```diff
@@ -16,59 +16,68 @@
...
-                 {extent3Minus, extent3Plus}}}) {}
+Extents::Extents(ast::cartesian_, int extent1minus, int extent1plus, int extent2minus,
+                 int extent2plus, int extent3minus, int extent3plus)
+    : horizontalExtent_(ast::cartesian_{}) {
+  horizontalExtent_ =
+      HorizontalExtent(ast::cartesian_{}, extent1minus, extent1plus, extent2minus, extent2plus);
+  verticalExtent_ = Extent(extent3minus, extent3plus);
+}
+
+Extents::Extents(ast::cartesian_) : horizontalExtent_(ast::cartesian_{}) {}
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/Extents.cpp]: reminder: `ast::cartesian`instead `ast::cartesian_{}`

File: dawn/src/dawn/IIR/Extents.cpp [hunk 14]
```diff
@@ -16,59 +16,68 @@
...
+
+Extents::Extents(const Extents& other)
+    : verticalExtent_(other.verticalExtent()), horizontalExtent_(other.horizontalExtent()) {}
 
-void Extents::addCenter(const unsigned int dim) {
-  DAWN_ASSERT(dim < 3);
+Extents::Extents(Extents&& other)
+    : verticalExtent_(other.verticalExtent()), horizontalExtent_(other.horizontalExtent()) {}
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/Extents.cpp]: horizontalExtent_(std::move(other.horizontalExtent()))

File: dawn/src/dawn/IIR/Extents.cpp [hunk 15]
```diff
@@ -16,59 +16,68 @@
...
-  extents_[dim].Plus = std::max(0, extents_[dim].Plus);
+Extents& Extents::operator=(const Extents& other) {
+  verticalExtent_ = other.verticalExtent();
+  horizontalExtent_ = other.horizontalExtent();
+  return *this;
+}
+
+Extents& Extents::operator=(Extents&& other) {
+  // TODO is this right?
+  verticalExtent_ = other.verticalExtent();
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/Extents.cpp]: also here: move from the RHS.

File: dawn/src/dawn/IIR/Extents.cpp [hunk 16]
```diff
@@ -16,59 +16,68 @@
...
+    : horizontalExtent_(ast::cartesian_{}) {
+  horizontalExtent_ =
+      HorizontalExtent(ast::cartesian_{}, extent1minus, extent1plus, extent2minus, extent2plus);
+  verticalExtent_ = Extent(extent3minus, extent3plus);
+}
+
+Extents::Extents(ast::cartesian_) : horizontalExtent_(ast::cartesian_{}) {}
+
+Extents::Extents(const Extents& other)
+    : verticalExtent_(other.verticalExtent()), horizontalExtent_(other.horizontalExtent()) {}
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/Extents.cpp]: alternative:
Extents::Extents(Extents const& other) {
    *this = other;
}
Extents::Extents(Extents&& other) {
    *this = std::move(other);
}

File: dawn/src/dawn/IIR/Extents.h [hunk 17]
```diff
@@ -34,7 +34,7 @@ namespace iir {
 
 /// @brief Access extent of a single dimension
 /// @ingroup optimizer
-struct Extent {
+struct Extent { // TODO class?
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/Extents.h]: what about this TODO?

File: dawn/src/dawn/IIR/Extents.h [hunk 18]
```diff
@@ -91,9 +94,158 @@ struct Extent {
...
+  void merge(HorizontalExtentImpl const& other) { merge_impl(other); }
+  void merge(dawn::ast::HorizontalOffset const& other) { merge_impl(other); }
+  void expand(HorizontalExtentImpl const& other) { expand_impl(other); }
+  void add(HorizontalExtentImpl const& other) { add_impl(other); }
+  bool equals(HorizontalExtentImpl const& other) const { return equals_impl(other); }
+  bool isPointwise() const { return isPointwise_impl(); }
+
+protected:
+  virtual void add_impl(HorizontalExtentImpl const& other) = 0;
+  virtual HorizontalExtentImpl* plus_impl(HorizontalExtentImpl const& other) const = 0;
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/Extents.h]: also this one should now return unique_ptr

File: dawn/src/dawn/IIR/FieldAccessExtents.h [hunk 19]
```diff
@@ -69,7 +68,7 @@ class FieldAccessExtents {
 
   std::optional<Extents> readAccessExtents_;
   std::optional<Extents> writeAccessExtents_;
-  Extents totalExtents_;
+  std::optional<Extents> totalExtents_;
```
Comment (lukasm91) [inline dawn/src/dawn/IIR/FieldAccessExtents.h]: why this change?  Might it be that totalExtents is always given, while readAccess and writeAccess are really optional?

File: dawn/test/unit-test/dawn/IIR/TestExtent.cpp [hunk 20]
```diff
@@ -19,116 +19,122 @@ using namespace dawn;
...
-  EXPECT_TRUE((extents[0] == Extent{-4, 4}));
-  EXPECT_TRUE((extents[1] == Extent{0, 0}));
-  EXPECT_TRUE((extents[2] == Extent{0, 0}));
+  compareExtents(addedExtents, {-4, 4, 0, 0, 0, 0});
 }
 
 TEST(ExtentsTest, Add5) {
-  Extents extents({-2, 2, 0, 4, 0, 0});
-  extents.add(ast::Offsets{ast::cartesian, 2, 2, 3});
+  // Extents extents({-2, 2, 0, 4, 0, 0});
```
Comment (lukasm91) [inline dawn/test/unit-test/dawn/IIR/TestExtent.cpp]: what about all these commented out tests?

Comment (mroethlin) [reply dawn/test/unit-test/dawn/IIR/TestExtent.cpp]: adding an offset is _only_ used by these tests but nowere else

File: dawn/src/dawn/IIR/FieldAccessExtents.h (hunk 19 above)
Comment (lukasm91) [reply dawn/src/dawn/IIR/FieldAccessExtents.h]: @mroethlin Can you help me here and tell me what you intend with this change?

File: dawn/src/dawn/IIR/Extents.h (hunk 17 above)
Comment (lukasm91) [reply dawn/src/dawn/IIR/Extents.h]: @mroethlin  Can you help me here? What is the state of this?

File: dawn/src/dawn/IIR/FieldAccessExtents.h (hunk 19 above)
Comment (havogt) [reply dawn/src/dawn/IIR/FieldAccessExtents.h]: The problem was that `Extents` doesn't have a default constructor, before it was constructed with zeros.

Comment (lukasm91) [reply dawn/src/dawn/IIR/FieldAccessExtents.h]: thanks. I put it back for the moment, because this causes segfaults / the optionals are not initialized when being accessed. We can put a proper fix when adding unstructured extents.

File: dawn/test/unit-test/dawn/IIR/TestExtent.cpp (hunk 20 above)
Comment (lukasm91) [reply dawn/test/unit-test/dawn/IIR/TestExtent.cpp]: will remove it completely

**Comment by lukasm91** [comment]
launch jenkins

**Comment by lukasm91** [comment]
launch jenkins

**Comment by lukasm91** [comment]
launch jenkins

**Comment by lukasm91** [comment]
launch jenkins

**Comment by lukasm91** [comment]
launch jenkins

**Comment by lukasm91** [comment]
launch jenkins
